man/tuxcast-config.1
.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH tuxcast-config 1 "MAY 2008" 0.2 "tuxcast 0.2"
.SH NAME
tuxcast-config \- tuxcast configuration utility
.SH SYNOPSIS
.B tuxcast-config [option] name
.SH DESCRIPTION
A configuration utility for Tuxcast. It must be run before the
main tuxcast program can be run. It generates the required XML
configuration files to define feeds and destination directories.

.SH OPTIONS
.IP -a -n name -A address -f folder
Add a feed named "name", with address "address".  Podcasts will be saved in the folder named "folder", under the root directory "podcastdir".  If folder is left blank, podcasts will be saved in the root podcast directory
.IP -d name
Delete the feed named "name".
.IP -g option
Retrieve the value of the option named "option"
.IP -g feed -n name
Show information on the feed named "name"
.IP -G
Show the value of all options, as well as listing all information on all feeds
.IP -h
Display a help message
.IP -s option=value
Set the option named "option" to "value"
.IP -m
List MIME types which will be downloaded.  If this list is empty, files of all MIME types will be downloaded
.IP -t type
Permit the MIME type "type" to be downloaded
.IP -r type
Remove the MIME type "type" from the list of permitted MIME types
.IP -u -n name [ -A address | -f folder | -N newname ]
Change a property of an existing feed.  You can only change one property at a time.  Use -A to set the address to "address", -f to set the folder in which to save podcasts to "folder", and -N to change the name of the feed to "newname"

Available options are:
* podcastdir - the root directory in which all podcasts are saved.
* ask - if set to "yes", the user will be asked before downloading each file

.SH BUGS
There are no known issues in this version of tuxcast-config
.SH AUTHOR
Tuxcast is written by David 'Cillian' Turner with occasional assistance from Roger 'Oojah' Light.  See the manpage for tuxcast for more details.

These manpages are by cypher_ of Linux User Group Radio (#lugradio) on the freenode network.
.SH "SEE ALSO"
.BR tuxcast (1),


CHANGELOG
Revision 171:
###############################################################################
===============================================================================
------------------------- Tuxcast 0.3 Released! -------------------------------
----------------------------- Luke warm tea -----------------------------------
===============================================================================
###############################################################################
Actual source tarballs, binaries and stuff will be up when I get around to it.
Oh, and package control files will be updated in here after the binaries are up

Revision 170:
* Updated TODO
* Updated manpages
--------------------------------------------------------------------------------

Revision 169:
* Updated TODO
* Now installs manpages
* PCRE now not required by default
--------------------------------------------------------------------------------

Revision 168:
* Now saves downloaded torrents in correct folders
--------------------------------------------------------------------------------

Revision 167:
* Removed a couple of debug messages
--------------------------------------------------------------------------------

Revision 166:
* Tuxcast now downloads more than the first torrent
--------------------------------------------------------------------------------

Revision 165:
* Added compile flag info to version output
* Updated TODO R.e. bittorrents
--------------------------------------------------------------------------------

Revision 164:
* Basic bittorrenting is *working* - Now it just needs to work with more than
	one file a session, among other things
--------------------------------------------------------------------------------

Revision 163:
* Imported libtorrent's example program as a library.
* Updated make system to facilitate optional torrent support
--------------------------------------------------------------------------------

Revision 162:
* Added the skeleton of a bittorrent handler subroutine (Possibly replace the
	current detection method with libmagic in future)
--------------------------------------------------------------------------------

Revision 161:
* Changed a bunch of stuff so tuxcast downloads all files after all feeds have
	been checked (or up2date'ed). (Prelude to bittorrent support)
--------------------------------------------------------------------------------

Revision 160:
* Replaced mktemp with mkstemp
--------------------------------------------------------------------------------

Revision 159:
* Updated TODO
* Updated compile flags
* Updated copyright dates
--------------------------------------------------------------------------------

Revision 158:
* Caught the eTuxcast_FSFull exception
* Updated TODO
* Changed the vanilla exception header to use fprintf and be translateable
--------------------------------------------------------------------------------

Revision 157:
* Actually added tuxcast_exceptions.h to SVN - forgot this in last rev.
--------------------------------------------------------------------------------

Revision 156:
* Added the move() call to filestuff
* Made tuxcast write config.xml and files.xml to temp files, then move the
	temp over the original - now if it runs out of space, it doesn't
	wipe the config anymore.  Fixes 1609639.
--------------------------------------------------------------------------------

Revision 155:
* Reverted out the xmas mode
* Added in optional PCRE support (The PCREPP define in compile_flags.h)
* Added in the configuration options to run commands after certain actions
	complete.  (The meat of the code isn't implemented yet)
* Removed some old eclipse metadata
--------------------------------------------------------------------------------

Revision 154:
* Added in the hidden "xmas" mode....
--------------------------------------------------------------------------------

Revision 153:
* Changed get() to mark a file as downloaded if the MIME type is wrong, so you
	don't keep seeing the message over and over again...
--------------------------------------------------------------------------------

Revision 152:
* Finished off MIME type filtering (I.e. implemented the tuxcast bit).
--------------------------------------------------------------------------------

Revision 151:
* Added options to tuxcast-config to add and remove permitted MIME types
--------------------------------------------------------------------------------

Revision 150:
* Added libxml includes to cleaner.cpp - somehow these disappeared :S
--------------------------------------------------------------------------------

Revision 149:
* Applied a patch from greatred to add in atom parsing and generally clean up
	RSS and RDF parsing
--------------------------------------------------------------------------------

Revision 148:
* Applied yet another patch from Greated - this one changes a bunch of stuff
	all over the place.  It's all internal code changes, there shouldn't
	be any noticable different on the outside.  Thanks again.
--------------------------------------------------------------------------------

Revision 147:
* Applied a second patch from Greated to make gettext string updating a *lot*
	neater - thanks again :).
--------------------------------------------------------------------------------

Revision 146:
* Applied a patch from Greatred to fix pluralisation of strings using gettext
	and to use ?: for tidying some bits of code.  Thanks :)
--------------------------------------------------------------------------------

Revision 145:
* Fixed a bunch of messy optiony errors in tuxcast.cpp caused by the merge going
	evilish. (i.e. s/optarg/optarg1/g)
--------------------------------------------------------------------------------

Revision 144:
* Removed an annoying and questionably useful warning message from the config
	loader
--------------------------------------------------------------------------------


config/tuxcast-config.cpp
/*
 * 
 * This file is part of Tuxcast, "The linux podcatcher"
 * Copyright (C) 2006-2007 David Turner
 * 
 * Tuxcast is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuxcast is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Tuxcast; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 * 
 */



#include "../compile_flags.h"
#include "config.h"
#include <unistd.h>
#include "config_exceptions.h"
#include "../libraries/filestuff_exceptions.h"
#include "../libraries/common.h"
#include "../version.h"
#include <libintl.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>

const char options[] = "aA:d:n:N:hs:g:Gf:umvt:r:";
#define _(x) gettext(x)

configuration myconfig;

void help(void);
void get(int argc, char **argv);
void getall(void);
void set();
void add(int argc,char *argc[]);
void del();
void update(int argc, char *argv[]);
void listmimes(void);
void addmime();
void removemime();

int main(int argc, char *argv[])
{
	// TODO: Insert internationalisation setup shizzle here:
	setlocale(LC_ALL,"");
	bindtextdomain("tuxcast","/usr/share/locale");
	textdomain("tuxcast");
	// TODO: Shouldn't that be tuxcast-config for the text domain?

	try
	{
		myconfig.load();
	}
	catch(eConfig_NoConfigFile &e)
	{
		fprintf(stderr,_("Warning, no config file\n"));
		fprintf(stderr,_("If you're changing an option or adding a feed, a config file will be created later\n"));
	}
	
	try
	{
		switch(getopt(argc,argv,options))
		{
			case 'a':
				add(argc,argv);
				break;
			
			case 'd':
				del();
				break;
			
			case 'g':
				get(argc,argv); // Get needs to use the options
				// for a few things.
				break;

			case 'G':
				getall();
				break;

			case 's':
				set();
				break;
			case 'u':
				update(argc,argv);
				break;

			case 'm':
				listmimes();
				break;

			case 't':
				addmime(); // Nothing is passed as the function
				// can access optarg...
				break;

			case 'r':
				removemime(); // "
				break;

			case 'v':
				version();
				break;

			case 'h':
				// Fallthrough
		
			default:
				help();
			}
	}
	catch(eConfig_CannotSaveConfig &e)
	{
		fprintf(stderr,_("Cannot save config file\n"));
		fprintf(stderr,_("Exception caught: "));
		e.print();
		fprintf(stderr,"\n");
		return -1;
	}
	// No point in catching that exception separatly in every action
	// Just catch here to avoid SIGABRT and exit cleanly
		
	return 0;
}

void listmimes(void)
{
	printf(_("Permitted MIME types:\n"));
	FOREACH(configuration::mimelist::iterator, myconfig.permitted_mimes, mime)
	{
		printf("%s\n",mime->c_str()); // Don't
		// add %s to the po file...
	}
}

void addmime(void)
{
	if(optarg == NULL)
	{
		fprintf(stderr,_("Optarg is NULL... this shouldn't happen.\n"));
		return; // TODO: exception?
	}
	FOREACH(configuration::mimelist::iterator, myconfig.permitted_mimes, mime)
	{
		if(strcasecmp(mime->c_str(),optarg) == 0)
		{
			fprintf(stderr,_("\"%s\" is already permitted\n"),optarg);
			return;
		}
	}

	myconfig.permitted_mimes.push_back(optarg);
	printf(_("\"%s\" added to list of permitted MIME types\n"),optarg);
	
	myconfig.save();
}

void removemime(void)
{
	if(optarg == NULL)
	{
		fprintf(stderr,_("Optarg is NULL... this shouldn't happen.\n"));
		return; // "
	}
	
	FOREACH(configuration::mimelist::iterator, myconfig.permitted_mimes, mime)
	{
		if(strcasecmp(mime->c_str(),optarg) == 0)
		{
			printf(_("\"%s\" is no longer permitted\n"),optarg);
			myconfig.permitted_mimes.erase(mime);
			myconfig.save();
			return;
		}
	}
	printf(_("\"%s\" was not found in the list of permitted MIME types\n"),optarg);

}

void help(void)
{

	printf(_("Usage: tuxcast-config  [action]\n"));
	printf(_("Actions are:\n"));
	printf(_("-a -n NAME -A ADDRESS -f FOLDER: Add a feed with the specified name, address and folder\n"));
	printf(_("-d NAME: Delete the feed with the specified name\n"));
	printf(_("-g OPTION: Get the value of the specified option\n"));
	printf(_("-g feed -n NAME: Get the settings of the specified feed\n"));
	printf(_("-G: Get the value of all options\n"));
	printf(_("-h: Display this help message\n"));
	printf(_("-s: OPTION=VALUE: Set OPTION to VALUE\n"));
	printf(_("-m: List permitted MIME types\n"));
	printf(_("-t TYPE: Allow a new mime type\n"));
	printf(_("-r TYPE: Reject a mime type\n"));
	printf(_("-u -n NAME [ -A ADDRESS | -f FOLDER | -N NEWNAME ]: Change a property of a feed\n\n"));
	printf(_("Available options are:\n"));
	printf(_("podcastdir - where to save all your podcasts\n"));
	printf(_("ask - whether or not to ask before downloading any files\n"));
}

void get(int argc, char **argv)
{
	string args = optarg; // This is the first option's optarg - although
	// getopt was called ageeees ago we can still retrieve optarg here.

	if(strcasecmp(args.c_str(),"podcastdir") == 0)
	{
		printf("%s\n",myconfig.podcastdir.c_str()); // No translation
		// required for just a format string, I spose.
		return;
	}
	if(strcasecmp(args.c_str(),"ask") == 0)
	{
		if(myconfig.ask == true)
			printf(_("true"));
		else
			printf(_("false"));
		return;
	}
	// DANGER DANGER WILL ROBINSON!!1  MY ARMS ARE FLAILING WILDLY!!!11one
	// If the thing after -g is shorter than 4 chars we could have a problem
	if(strcasecmp(args.c_str(),"feed") == 0)
	{
		if(getopt(argc,argv,options) != 'n')
		{
			fprintf(stderr,_("You must pass a feed name with -n NAME\n"));
			return;
		}
		string name = optarg;
		// Locate which feed they are talking about:
		FOREACH(configuration::feedlist::iterator, myconfig.feeds, feed)
		{
			if(strcasecmp(feed->name.c_str(),name.c_str()) == 0) // Found it:
			{
				feed->displayConfig();
				return;
			}
			else
				continue; // RRRAAAWWWRRRR.  :P
		}
		// If we get here the feed name is bollocks:
		fprintf(stderr,_("Invalid feed name, %s\n"),name.c_str());
		return;
	}
	if(strcasecmp(args.c_str(),"feeds") == 0)
	{
		// Display all feeds:
		FOREACH(configuration::feedlist::iterator, myconfig.feeds, feed)
		{
			feed->displayConfig();
		}
		return;

	}
	// other variables go here...
	
	// If we are still here and haven't returned, somebody messed up.
	// Probably the user ;)
	fprintf(stderr,_("Unrecognised option, %s\n"),args.c_str());
}

void getall(void)
{
	printf(_("podcastdir = %s\n"),myconfig.podcastdir.c_str());
	printf(_("ask = %s\n"), (myconfig.ask?"true":"false"));
	listmimes();
	// ...
	// Let's show some feeds:
	printf(ngettext("There is 1 feed:\n", "There are %d feeds:\n",
				myconfig.feeds.size()),myconfig.feeds.size());
	FOREACH(configuration::feedlist::iterator, myconfig.feeds, feed)
	{
		feed->displayConfig();
	}

}

void set()
{
	string args = optarg;
	string varname=args.substr(0,args.find("=",0));
	string value=args.substr(args.find("=",0)+1,args.length()-args.find("=",0)-1);

	if(strcasecmp(varname.c_str(),"podcastdir") == 0)
	{
		myconfig.podcastdir = value;
	}
	if(strcasecmp(varname.c_str(),"ask") == 0)
	{
		if((strcasecmp(value.c_str(),"true") == 0) || (strcasecmp(value.c_str(),"yes") == 0))
		{
			myconfig.ask = true;
		}
		else
		{
			if((strcasecmp(value.c_str(),"false") == 0) || (strcasecmp(value.c_str(),"no") == 0))
			{
				myconfig.ask = false;
			}
			else
			{
				fprintf(stderr,_("The \"ask\" variable can only be \"true\" or \"false\" -\nAssuming true\n"));
				myconfig.ask = true;
			}
		}
	}
	// ...
	
	myconfig.save();
}

void add(int argc,char *argv[])
{
	int myopt=0;
	string name, address, folder;

	while((myopt = getopt(argc,argv,options)) != -1)
	{
		if((char)myopt == 'A')
			address = optarg;
		if((char)myopt == 'n')
			name = optarg;
		if((char)myopt == 'f')
			folder = optarg;
	}

	if(name.empty())
	{
		fprintf(stderr,_("No name specified!\n"));
		return;
	}
	if(address.empty())
	{
		fprintf(stderr, _("No URL specified!\n"));
		return;
	}
	// folder can be blank - it means just put the podcast in podcastdir

	FOREACH(configuration::feedlist::iterator, myconfig.feeds, feed)
	{
		// Let's check the feed isn't already there:
		if(strcasecmp(feed->name.c_str(),name.c_str()) == 0)
		{
			fprintf(stderr,_("Feed already exists\n"));
			return;
		}
		// Check the URL isn't a duplicate:
		if(strcasecmp(feed->address.c_str(),address.c_str()) == 0)
		{
			fprintf(stderr, _("A feed with the same URL already exists\n"));
			return;
		}
	}

	myconfig.feeds.push_back(feed(name, address, folder));
	
	myconfig.save();
}

void del()
{
	FOREACH(configuration::feedlist::iterator, myconfig.feeds, feed)
	{
		if(strcasecmp(optarg,feed->name.c_str()) == 0)
		{
			// We've found the feed - wipe it:
			myconfig.feeds.erase(feed);
			myconfig.save();
			return;
		}
	}
	fprintf(stderr,_("Feed \"%s\" doesn't exist\n"),optarg);
}
	
void update(int argc, char *argv[])
{
	char myopt;
	string name;
	// The -u option was already sucked up
	// The next option must be -n so we know what to update
	// After that, we need either -f or -A
	
	myopt = getopt(argc,argv,options);
	if(myopt != 'n')
	{
		fprintf(stderr,_("You must pass -n and either -A, -f or -N when updating a feed's info\n"));
		return;
	}
	name = optarg;
	myopt = getopt(argc,argv,options);
		
	FOREACH(configuration::feedlist::iterator, myconfig.feeds, feed)
	{
		if(strcasecmp(feed->name.c_str(),name.c_str()) == 0)
		{
			// Found the feed, update it:
			switch(myopt)
			{
				case 'A':
					if(strcmp(optarg,"") == 0)
					{
						fprintf(stderr,_("Error: Blank address\n"));
						return;
					}
					FOREACH(configuration::feedlist::iterator, myconfig.feeds, feed_finder)
					{
						if(strcmp(feed_finder->address.c_str(), optarg) == 0)
						{
							fprintf(stderr,_("Error: Another feed already exists with this address\n"));
							return;
						}
					}
					
					feed->address = optarg;
					break;

				case 'f':
					feed->folder = optarg;
					break;
					
				case 'N':
					if(strcmp(optarg,"") == 0)
					{
						fprintf(stderr,_("Error: Blank name\n"));
						return;
					}
					
					FOREACH(configuration::feedlist::iterator, myconfig.feeds, feed_finder)
					{
						if(strcmp(feed_finder->name.c_str(), optarg) == 0)
						{
							fprintf(stderr,_("Error: Another feed already exists with this name\n"));
							return;
						}
					}
					
					feed->name=optarg;
					break;
					
				default:
					fprintf(stderr,_("You must pass either -A, -f or -N when updating a feed's info\n"));
					return;
			}

			// We've found the feed and updated it
			myconfig.save();
			return;
		}
	}
fprintf(stderr,_("Error: feed \"%s\" doesn't exist.\n"),name.c_str());
return;
}
